### ISSUE TITLE
Errors from use of pyright with pydantic's RootModel

### ISSUE THREAD

**Issue by acampove** [microsoft/pyright#11319]
**Describe the bug**
I am type annotating properly a `RootModel` class and I am still getting an error. The pydantic maintainers told me I should report it to the pyright maintainers:

https://github.com/pydantic/pydantic/issues/12899#issuecomment-4012461405

**Code or Screenshots**

```python
from pydantic import RootModel

class MVAWp(RootModel[float | tuple[float,float]]):
    def __str__(self):
        if isinstance(self.root, float):
            return f'{self.root:.3f}'

        low, high = self.root
        return f'{low:.3f}-{high:.3f}'

val = MVAWp(3)
print(val)

val = MVAWp((1, 2))
print(val)
```

**Comment by Viicos** [comment]
To clarify, this has nothing to do with Pydantic. Here is a MRE:

```python
def func(a: float):
    if isinstance(a, float):
        return

    reveal_type(a)  # int
```

@acampove I would suggest looking for duplicates before opening issues. This has already been reported in https://github.com/microsoft/pyright/issues/11114.

**Comment by erictraut** [comment]
This behavior is correct, not a bug. See #11114 for details.

**Comment by acampove** [comment]
Hi @Viicos @erictraut 

That's not a reproducer of this problem. The problem happens if I put an int or a float in the code:

```python
from pydantic import RootModel

class MVAWp(RootModel[float | tuple[float,float]]):
    def __str__(self):
        if isinstance(self.root, float):
            return f'{self.root:.3f}'

        low, high = self.root
        return f'{low:.3f}-{high:.3f}'

val = MVAWp(3.)
print(val)

val = MVAWp((1., 2.))
print(val)
```

THAT is a reproducer. I still see the problem:

<img width="1379" height="420" alt="Image" src="https://github.com/user-attachments/assets/2ed73be4-e6e7-4c0c-9d84-3987c0e19d5b" />

I am giving it a float, it sees an int, how is that a problem in my code? And, just to be even more explicit:

```python
from pydantic import RootModel

class MVAWp(RootModel[float | tuple[float,float]]):
    def __str__(self):
        if isinstance(self.root, float):
            return f'{self.root:.3f}'

        low, high = self.root
        return f'{low:.3f}-{high:.3f}'

fval : float = 3.

val = MVAWp(fval)
print(val)

val = MVAWp((fval, fval))
print(val)
```

**Comment by erictraut** [comment]
@acampove, did you read the explanation in #11114? Does it make sense?

**Comment by acampove** [comment]
> [@acampove](https://github.com/acampove), did you read the explanation in [#11114](https://github.com/microsoft/pyright/issues/11114)? Does it make sense?

I read it, I did not take this

> the type annotation float implies float | int

seriously. I knew something did not make sense and some bad design had to be hidding somewhere, pyright, pydantic or python itself.
